### DISCUSSION TITLE
What are subfields & how can my custom field support it?

### DISCUSSION THREAD

**Discussion by ziming** [Laravel-Backpack/community-forum#608] [Q&A (Help)]
Hi, recently I got an issue on my package repo saying that my field does not work on subfield

https://github.com/ziming/json-field-for-backpack/issues/10

However I have never used subfield on backpack before. When i try to search backpack documentation on subfield, it brings me to uploader, repeater and I'm kind of loss on what is subfield about & what special things I need to do to support my field with it

So I decide to ask my question here so I can fix my package

**Comment by pxpm** [comment]
Hey @ziming happy to help. 

Subfields are exactly what you have found, they are used inside the "repeatable containers" (Solo, to store as json inside a column, or some relationships like HasOne, BelongsToMany and HasMany (including the Morphs) also support the repeatable interface. They have the same definition as the fields eg:
```php
CRUD::field('container')
  ->type('repeatable')
  ->subfields([
    [
      'name' => 'my_subfield',
      'type' => 'your_json_field_type'
    ],
  ]);
```

Since your field is a `json` field, what I suspect that may happen is that we mess it up when `encoding/decoding` the repeatable json. 

If you can provide me an error I can try to find if the issue is on our side, or if we can do something to make it work properly. 

Cheers

**Reply by ziming** [reply]
Hmm the error is by the person who created an issue on my repo, let me ask them for an example

The error message they had is Uncaught ReferenceError: JSONEditor is not defined

```php
CRUD::field([ 'name'  => 'test', 
'type'          => "relationship",
 'tab'           => 'Test',
 'label'         => 'Test',
 'subfields'   => [ 
[ 'name' => 'test_relationship_id',
 'label'      => 'Json textarea',
 'type'       => 'json',
 'view_namespace' => 'json-field-for-backpack::fields', 
'modes' => ['form', 'tree', 'code'], 
'tab'        => 'Test', 
'wrapper' => [ 'class' => 'form-group col-md-6', 
], 
], 
], 
]);
```

**Reply by pxpm** [reply]
I will try your package and let you know if I find something broken. 

It could be an issue the way you load the jsoneditor script, it needs to probably be available globally and earlier in the lifecycle of the application. 

I can't promise today, but in the next couple days I will give it a try 🙏

**Reply by tabacitu** [reply]
@pxpm this would be a good opportunity for us to create some sort of docs... or pointers... or rules... for people who develop custom Fields. Or at least a note at the end of the Create Custom Field docs. Something like...

> If you want your field to work inside `repeatable` and `relationship` too, as a subfield, then please:
> 1. xxx
> 2. yyy
> 3. zzz
